packages/patrol/lib/src/custom_finders/patrol_finder.dart

```diff
@@ -306,13 +306,25 @@ class PatrolFinder extends MatchFinder {
   /// * [CommonFinders.byWidgetPredicate]
   PatrolFinder which<T extends Widget>(bool Function(T widget) predicate) {
     return PatrolFinder(
-      finder: find.byWidgetPredicate((widget) {
-        if (widget is T) {
-          return predicate(widget);
-        } else {
-          return false;
-        }
-      }),
+      finder: find.descendant(
+        matchRoot: true,
+        of: finder,
+        matching: find.byWidgetPredicate((widget) {
+          if (widget is T &&
+              find
+                  .descendant(of: finder, matching: find.byWidget(widget))
+                  .evaluate()
+                  .isEmpty &&
+              find
+                  .ancestor(of: finder, matching: find.byWidget(widget))
+                  .evaluate()
+                  .isEmpty) {
```

Comment:
Hmm, could we make this `if` more readable? Perhaps by splitting it into `final`s and then doing:

```dart
final nicelyNamedCondition1 = ...;
final nicelyNamedCondition2 = ...;
if (nicelyNamedCondition1 && nicelyNamedCondition2) {
 ...
}
```

Reply:
Also, maybe we can reduce the amount of work being done by calling `find.byWidget(widget)` once?